core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java

```diff
@@ -572,6 +556,19 @@ public static boolean assertNotClusterStateUpdateThread(String reason) {
         return true;
     }
 
+    /** asserts that the current stack trace does <b>NOT</b> invlove a cluster state applier */
+    private static boolean assertNotCalledFromClusterStateApplier(String reason) {
```

Comment:
OK. I've done it but I had to revert it back. The problem is that there is a [validator that samples the state](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java#L165). I think this is a good idea and we should address it but I rather not tackle it now.